src/components/album/ListeningAlbum.js

```diff
@@ -0,0 +1,70 @@
+// Import necessary modules from React
+import React, { useEffect } from 'react';
+import io from 'socket.io-client';
+import { useAuth } from '@akosasante/react-auth-context';
+
+// Define the ListeningAlbum component
+const ListeningAlbum = ({ albumId }) => {
+  const { user } = useAuth();
+
+  useEffect(() => {
+    // Create a socket connection to your backend
+    const socket = io(process.env.REACT_APP_SOCKET_BASE_PATH); // Replace with your actual backend URL and port
+    socket.on('connect', () => {
+      console.log('Connected to WebSocket server');
+    });
+    // Function to handle play button click
+    const handlePlayClick = () => {
+      const userId = user?._id;
+      socket.emit('listening-to-album-play', { albumId, userId });
+    };
+
+    // Function to handle pause button click
+    const handlePauseClick = () => {
+      const userId = user?._id;
+      socket.emit('listening-to-album-pause', { albumId, userId });
+    };
+
+    // Function to handle page unload
+    const handleBeforeUnload = () => {
+      socket.emit('user-leaving-page');
+    };
+
+    // Add event listeners when the component mounts
+    document
+      .getElementById('playButton')
+      ?.addEventListener('click', handlePlayClick);
+    document
+      .getElementById('pauseButton')
+      ?.addEventListener('click', handlePauseClick);
+    window.addEventListener('beforeunload', handleBeforeUnload);
```

Comment:
While these will work in this case since the component is fairly simple; it may become problematic as we expand on the functionality. Generally, it is best practice to _not_ use `addEventListener` in React code. This is because those interact with the DOM directly. Which means when React is re-rendering this component (for example if a state variable or path variable changes) then it may be re-adding these event listeners in duplicate. In this case it's unlikely since this `useEffect` is only rerun if the `user id` or `albumId` change. But it's still best to stick to React's own listeners when possible. 

https://legacy.reactjs.org/docs/handling-events.html#:~:text=When%20using%20React%2C%20you%20generally%20don%E2%80%99t%20need%20to%20call%20addEventListener%20to%20add%20listeners%20to%20a%20DOM%20element%20after%20it%20is%20created.%20Instead%2C%20just%20provide%20a%20listener%20when%20the%20element%20is%20initially%20rendered.

For `click` listeners, this would mean adding `onClick` attribute to the buttons. Eg:

```javascript
// move handler to outside of the `useEffect`

// Function to handle play button click
const handlePlayClick = () => {
  console.log('playing')
  const userId = user?._id;
  socket.emit('listening-to-album-play', { albumId, userId });
};

// In the render (return statement) tell React to call the handler when the button element is clicked on
// Render the component
return (
  <>
    <h1>Listening to Album</h1>
    <button onClick={handlePlayClick} id="playButton">Play</button>
    <button id="pauseButton">Pause</button>
  </>
);
```

Same can be done for the pause button. Since these functions need access to the `socket` variable and we want to make sure we're using the same `socket` regardless of if React needs to re-render this component, we should store that in a `useState` variable. You can refer to `AlbumChatBox.js` for the approach used there.

For `handleBeforeUnload`, it sort of works. It's listening for a `beforeunload` DOM event. So if the other user closes their tab or window; or does a full refresh of the browser page, then it will fire. But if they just click on "Home" or "Profile" or "Sign Out" to navigate to a different page within the React app, then the `beforeunload` event **will not** occur. The more React-y way to handle this is to call this `handleBeforeUnload` function (or just write those lines in-line) in the `return` value of the `useEffect` so that they are called as clean up steps when the component is being unmounted. This will also work on the `Album.js` modal, so as soon as someone closes the modal, that will emit a "user has left the room" event.

Reply:
@akosasante i think this is it, //still trying to figure our how it works, will now study the videos/refs you've left in this ticket https://codethedreamteam4-2023.atlassian.net/browse/PS-187